Add limited cursor support (left / right)

icamaster opened the pull request:
This adds limited cursor support for cursor left / right. Let me know what you think.

I started working on this a few days ago, but just noticed that you have put this repo in maintenance mode. Does this mean no PRs for features will be accepted? Thanks!

funbiscuit commented:
Thanks for your contribution! PRs are always welcome, I'll look into this one when I have time. 
But I see that CI doesn't pass, please take a look at errors.

icamaster commented:
Thanks, I'll have a look at the failures later this week. Feel free to ignore it until I add the fixes.

icamaster commented:
Ok, I believe I have added all the necessary fixes now. 

As a side question, if I want to run the CI on my fork, do you think it would be better to just edit the workflows on my master branch, or add a 'staging' branch on my fork, and test my PR changes there before pushing here? The former sounds easier, but the latter sounds like the correct approach. Thanks!

funbiscuit commented on README.md:
```diff
@@ -14,6 +14,7 @@ Single-header CLI library intended for use in embedded systems (like STM32 or Ar
 * Live autocompletion (see demo above, can be disabled)
 * Tab (jump to end of current autocompletion) and backspace (remove char) support
 * History support (navigate with up and down keypress)
+* Limited cursor support
```
Why is it limited? I think it will be better to explain, for example (left and right keypress to navigate inside input)

icamaster replied:
Added suggested comment

funbiscuit commented on lib/src/embedded_cli.c:
```diff
@@ -194,6 +211,26 @@ static const uint16_t cliInternalBindingCount = 1;
 
 static const char *lineBreak = "\r\n";
 
+/* Reference for VT100 escape sequences: https://learn.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences */
```
Please also link [ECMA-48](https://ecma-international.org/publications-and-standards/standards/ecma-48)

icamaster replied:
Done

funbiscuit commented on lib/src/embedded_cli.c:
```diff
@@ -194,6 +211,26 @@ static const uint16_t cliInternalBindingCount = 1;
+/** Escape sequence - Cursor insert character */
+static const char *escSeqInsertChar = "\x1B[@";
+
+/** Escape sequence - Cursor delete character */
```
Add sequence code `DCH` to comment

icamaster replied:
Done

funbiscuit commented on lib/src/embedded_cli.c:
```diff
@@ -194,6 +211,26 @@ static const uint16_t cliInternalBindingCount = 1;
+/** Escape sequence - Cursor restore position */
+static const char *escSeqCursorRestore = "\x1B[u";
+
+/** Escape sequence - Cursor insert character */
```
Add sequence code to comment `ICH`

icamaster replied:
Done

funbiscuit reviewed (changes requested):
I finally found some time to look at code. I've added some questions and there is also a bug:

- Enter 'abcd'
- Press Left two times
- Press backspace five times
You'll see that command prompt `> ` was cleared and output is corrupted.

There are some other bugs but it looks like they will be fixed if this one is fixed (they all appear when left and backspace are used). Please take a look and fix it. Ideally, add a test case first, that reproduces the bug.

Also this change breaks arduino example, now it requires two more bytes to work. Please edit lines [23](https://github.com/funbiscuit/embedded-cli/blob/master/examples/arduino-cli/arduino-cli.ino#L23) and [24](https://github.com/funbiscuit/embedded-cli/blob/master/examples/arduino-cli/arduino-cli.ino#L24) to use 166 bytes instead of 164.

icamaster commented:
Thank you for looking at this PR. 


> I finally found some time to look at code. I've added some questions and there is also a bug:
> 
> * Enter 'abcd'
> * Press Left two times
> * Press backspace five times
>   You'll see that command prompt `> ` was cleared and output is corrupted.
> 
> There are some other bugs but it looks like they will be fixed if this one is fixed (they all appear when left and backspace are used). Please take a look and fix it. Ideally, add a test case first, that reproduces the bug.

I think I have fixed the bug you mentioned and added a few tests to check for it. 

Two other 'quirks' which I am not sure if they should be fixed or not are:

- Doing an autocomplete adds a whitespace. If you don't delete this whitespace, then further autocompletes won't work if you forget to delete it. The fix would be to trim the current input before autocompletion, but not sure if it needs to be done. For example, using the 'Windows' example which has a binding for 'hello' :
  - Type 'hell'
  - Press tab to autocomplete
  - Press left key to move cursor once then backspace to delete 'o'
  - Press tab to autocomplete again - this will fail because of the whitespace at the end

- Requesting an autocomplete in the middle of a command by pressing 'tab' doesn't bring the cursor to the end if the command is already fully typed. This is mainly because 'getAutocompletedCommand' returns 'candidateCount = 0' for a fully typed command. Should this be the expected behavior? For example, using the 'Windows' example which has a binding for 'hello' :
  - Type 'hello'
  - Press the left key two times to move the cursor to the left two spaces
  - Press tab to autocomplete - nothing happens

> Also this change breaks arduino example, now it requires two more bytes to work. Please edit lines [23](https://github.com/funbiscuit/embedded-cli/blob/master/examples/arduino-cli/arduino-cli.ino#L23) and [24](https://github.com/funbiscuit/embedded-cli/blob/master/examples/arduino-cli/arduino-cli.ino#L24) to use 166 bytes instead of 164.

Good spot, fixed this. I believe this could be improved in the checks by adding some 'static_assert's, but might require some rewrite and is out of the scope of this PR.

funbiscuit commented:
> Doing an autocomplete adds a whitespace.

Ideally, this should be fixed. If you have time, you can fix it in this PR. But it may be not as straight-forward as it might seem. For example input `hello    ` (there are 4 spaces) after two left keypresses and tab should still have all extra spaces to the right of the cursor. So maybe it's a good idea to do this in a separate PR (or never bother with it at all). I'm okay with merging this PR in current state - everything looks great.

> Requesting an autocomplete in the middle of a command by pressing 'tab' doesn't bring the cursor to the end if the command is already fully typed

I wasn't able to reproduce it. Can you check if following test will fail?

- Input `hello`
- Press left two times
- Press tab
- expect input to be `> hello` and cursor position to be 8

If I do this in linux example, everything works as expected.

icamaster commented:
Thank you for looking at this so soon. 

> > Doing an autocomplete adds a whitespace.
> 
> Ideally, this should be fixed. If you have time, you can fix it in this PR. But it may be not as straight-forward as it might seem. For example input `hello ` (there are 4 spaces) after two left keypresses and tab should still have all extra spaces to the right of the cursor. So maybe it's a good idea to do this in a separate PR (or never bother with it at all). I'm okay with merging this PR in current state - everything looks great.

Agree that it might be a bit complex. I think I will look at this in the future, but will keep it out of this PR for now. Happy to raise an issue against this if you end up merging this PR. 

> > Requesting an autocomplete in the middle of a command by pressing 'tab' doesn't bring the cursor to the end if the command is already fully typed
> 
> I wasn't able to reproduce it. Can you check if following test will fail?
> 
> * Input `hello`
> * Press left two times
> * Press tab
> * expect input to be `> hello` and cursor position to be 8
> 
> If I do this in linux example, everything works as expected.

You are correct, it is working as expected.

I looked again over my initial notes and realized that it was the same problem as described above, I just forgot to mention that I used autocomplete (tab) before moving the cursor left: 

* Input `hello`
* Press left two times
* Press tab
* Expect input to be `> hello` and cursor position to be 8
* Press left two times again
* Press tab again
* Expect nothing to happen because of the whitespace added by the first autocomplete
